### PR TITLE
Projectkk2glider/issue 2239 ppm pushpull  for 9xrpro

### LINKED ISSUE #2239: PPM modulation opendrive/fulldrive
Hello i update my new 9xrPro to last version of opentx but tne my module assan v1 not work!! The problem is the protocol of ppm modulation..where i can change modulation ppp fulldrive?there is no menu!! Thx

### PR DESCRIPTION
**Pull request by projectkk2glider**
This is an attempt to fix #2239.

@bsongis I want you to tell me if I am going in the right direction. I've first tried to add opendrain option to the existing line for PPM parameters (length, pause, polarity) but there isn't enough space. I find working with menus very hard! The MENU_TAB is very hard to read/edit.

When this is completed I will rebase and squash commits to make it into one commit.

### CODE REVIEW

**Comment by bsongis** [comment]
The MENU_TAB macro comes from what we did for 9X, it could be simplified for Taranis and other ARM boards, but it's not an easy change. I am open to any suggestions, but won't like if it uses too much RAM!!

File: radio/src/gui/9X/menu_model_setup.cpp [hunk 1]
```diff
@@ -125,7 +134,7 @@ void menuModelSetup(uint8_t event)
   #define EXTRA_MODULE_ROWS
 #endif
   #define TRAINER_MODULE_ROWS
-  MENU_TAB({ 0, 0, TIMER_ROWS, TIMER_ROWS, TIMER_ROWS, 0, 1, 0, 0, 0, 0, 0, CASE_CPUARM(LABEL(PreflightCheck)) CASE_CPUARM(0) 0, 6, NUM_STICKS+NUM_POTS+NUM_ROTARY_ENCODERS-1, 0, LABEL(ExternalModule), (IS_MODULE_XJT(EXTERNAL_MODULE) || IS_MODULE_DSM2(EXTERNAL_MODULE)) ? (uint8_t)1 : (uint8_t)0, EXTERNAL_MODULE_CHANNELS_ROWS(), (IS_MODULE_XJT(EXTERNAL_MODULE) && IS_D8_RX(EXTERNAL_MODULE)) ? (uint8_t)1 : (IS_MODULE_PPM(EXTERNAL_MODULE) || IS_MODULE_XJT(EXTERNAL_MODULE) || IS_MODULE_DSM2(EXTERNAL_MODULE)) ? (uint8_t)2 : HIDDEN_ROW, IF_EXTERNAL_MODULE_XJT(FAILSAFE_ROWS(EXTERNAL_MODULE)), EXTRA_MODULE_ROWS TRAINER_MODULE_ROWS });
+  MENU_TAB({ 0, 0, TIMER_ROWS, TIMER_ROWS, TIMER_ROWS, 0, 1, 0, 0, 0, 0, 0, CASE_CPUARM(LABEL(PreflightCheck)) CASE_CPUARM(0) 0, 6, NUM_STICKS+NUM_POTS+NUM_ROTARY_ENCODERS-1, 0, LABEL(ExternalModule), (IS_MODULE_XJT(EXTERNAL_MODULE) || IS_MODULE_DSM2(EXTERNAL_MODULE)) ? (uint8_t)1 : (uint8_t)0, EXTERNAL_MODULE_CHANNELS_ROWS(), EXTERNAL_MODULE_SETTINGS_ROWS(), OPENDRAIN_ROWS() IF_EXTERNAL_MODULE_XJT(FAILSAFE_ROWS(EXTERNAL_MODULE)), EXTRA_MODULE_ROWS TRAINER_MODULE_ROWS });
```
Comment (bsongis) [inline radio/src/gui/9X/menu_model_setup.cpp]: Yes, it is better like that!

File: radio/src/myeeprom.h [hunk 2]
```diff
@@ -414,7 +414,13 @@ PACK(typedef struct {
...
   int8_t  ppmDelay;
   int8_t  ppmFrameLength;
-  uint8_t ppmPulsePol;
+  uint8_t ppmPulsePol:1;
+#if defined(PCBSKY9X) && defined(REVX)
+  uint8_t ppmOpenDrain:1;
+  uint8_t spare:6;
+#else
+  uint8_t spare:7;
+#endif
```
Comment (bsongis) [inline radio/src/myeeprom.h]: I would even have saved one byte in this structure. But well, it avoids a new shift as you did, let's keep it like that!

**Comment by bsongis** [comment]
Yes it's excellent, just the companion part which is missing.

**Comment by projectkk2glider** [comment]
Well this is it. I've never played with eeprom changes, menus and new translations, so please verify carefully.

File: radio/src/gui/9X/menu_model_setup.cpp [hunk 3]
```diff
@@ -112,6 +115,12 @@ void menuModelSetup(uint8_t event)
   #define IF_EXTERNAL_MODULE_ON(x)          (g_model.externalModule == MODULE_TYPE_NONE ? HIDDEN_ROW : (uint8_t)(x))
   #define IS_D8_RX(x)                       (g_model.moduleData[x].rfProtocol == RF_PROTO_D8)
   #define EXTERNAL_MODULE_CHANNELS_ROWS()   IF_EXTERNAL_MODULE_ON(IS_MODULE_DSM2(EXTERNAL_MODULE) ? (uint8_t)0 : (uint8_t)1)
+  #define EXTERNAL_MODULE_SETTINGS_ROWS()   (IS_MODULE_XJT(EXTERNAL_MODULE) && IS_D8_RX(EXTERNAL_MODULE)) ? (uint8_t)1 : (IS_MODULE_PPM(EXTERNAL_MODULE) || IS_MODULE_XJT(EXTERNAL_MODULE) || IS_MODULE_DSM2(EXTERNAL_MODULE)) ? (uint8_t)2 : HIDDEN_ROW
+#if defined(PCBSKY9X) && defined(REVX)
+  #define OUTPUT_TYPE_ROWS()                  (IS_MODULE_PPM(EXTERNAL_MODULE) ? (uint8_t)0 : HIDDEN_ROW) ,
```
Comment (bsongis) [inline radio/src/gui/9X/menu_model_setup.cpp]: Just a bad indentation here

File: radio/src/translations.cpp [hunk 4]
```diff
@@ -150,11 +153,14 @@ const pm_char STR_TIMER[] PROGMEM = TR_TIMER;
...
 const pm_char STR_TRIMINC[] PROGMEM = TR_TRIMINC;
-const pm_char STR_DISPLAY_TRIMS[] PROGMEM = TR_DISPLAY_TRIMS;
+const pm_char STR_DISPLAY_TRIMS[] PROGMEM = TR_DISPLAY_TRIMS;
 const pm_char STR_TTRACE[] PROGMEM = TR_TTRACE;
 const pm_char STR_TTRIM[] PROGMEM = TR_TTRIM;
 const pm_char STR_BEEPCTR[] PROGMEM = TR_BEEPCTR;
-const pm_char STR_USE_GLOBAL_FUNCS[] PROGMEM = TR_USE_GLOBAL_FUNCS;
+const pm_char STR_USE_GLOBAL_FUNCS[] PROGMEM = TR_USE_GLOBAL_FUNCS;
+#if defined(PCBSKY9X) && defined(REVX)
+    const pm_char STR_OUTPUT_TYPE[] PROGMEM = TR_OUTPUT_TYPE;
```
Comment (bsongis) [inline radio/src/translations.cpp]: 2 chars indentation

**Comment by bsongis** [comment]
Excellent work, thanks!

**Comment by projectkk2glider** [comment]
Rebased and squashed into one commit.

**Pull request merged**